lib/full-site-editing/class-wp-rest-templates-controller.php

```diff
@@ -0,0 +1,540 @@
+<?php
+/**
+ * REST API: WP_REST_Templates_Controller class
+ *
+ * @package    Gutenberg
+ * @subpackage REST_API
+ */
+
+/**
+ * Base Templates REST API Controller.
+ */
+class WP_REST_Templates_Controller extends WP_REST_Controller {
+	/**
+	 * Post type.
+	 *
+	 * @var string
+	 */
+	protected $post_type;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string $post_type Post type.
+	 */
+	public function __construct( $post_type ) {
+		$this->post_type = $post_type;
+		$this->namespace = 'wp/v2';
+		$obj             = get_post_type_object( $post_type );
+		$this->rest_base = ! empty( $obj->rest_base ) ? $obj->rest_base : $obj->name;
+	}
+
+	/**
+	 * Registers the controllers routes.
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+		// Lists all templates.
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_items' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+					'args'                => $this->get_collection_params(),
+				),
+				array(
+					'methods'             => WP_REST_Server::CREATABLE,
+					'callback'            => array( $this, 'create_item' ),
+					'permission_callback' => array( $this, 'create_item_permissions_check' ),
+					'args'                => $this->get_endpoint_args_for_item_schema( WP_REST_Server::CREATABLE ),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+
+		// Lists/updates a single template based on the given id.
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/(?P<id>[\/|\w-]+)',
```

Comment:
While testing the site editor, I've noticed that, while the list requests work fine, the single request fails with `rest_not_found`.

It might be because the `|` is encoded in the URL, and the request is sent as, for example:
`wp/v2/templates/tt1-blocks%7Cfront-page`
Apparently, the regex only matches the first part of the ID (e.g. `tt1-blocks`).

I've tried again by naïvely updating the regex to `/(?P<id>[\/%7C\w-]+)`, and indeed it matches the full ID.

Although, it's then kept with the encoded `|` (`tt1-blocks%7Cfront-page`), which of course doesn't work, and the request returns `rest_template_not_found`.

Reply:
Did you try with the latest updates? What are the exact steps to reproduce errors?

Reply:
Yeah it's happening to me even after a fresh pull. 🤔 

No real steps: I've just opened the site editor, which loads as a white screen, due to the current template not being returned by the API.

<img width="1006" alt="Screenshot 2021-01-07 at 12 39 50" src="https://user-images.githubusercontent.com/2070010/103893627-7d02cb00-50e5-11eb-8ae0-492efd2dde69.png">

As you can see in the screen, the list requests work fine, while the single one can't find a matching route.

Other hints: right before this test, I've ran `npm run wp-env start -- --update` to update the Docker images, and activated TT1 Blocks for the first time (last time I was still using it as TwentyTwentyOne Blocks).

Reply:
I don't reproduce and the e2e tests actually do perform the action you're doing and they are working properly. I wonder if you have any old leftovers like auto-draft templates or things like that.

Reply:
Oh, I've found out what's causing my issue: the permalink settings.

Basically, my dev site has rewritten permalinks, such as `http://localhost:8888/2021/01/07/sample-post/`, which also affects the API paths, by rewriting them automatically as `/wp-json/wp/v2/` etc.

So, with "pretty" permalinks, the full request (which fails) is:
`http://localhost:8888/wp-json/wp/v2/templates/tt1-blocks%7Cfront-page`

If I revert to the default plain permalinks (e.g. `http://localhost:8888/?p=123`), the API paths are now passed as a `?rest_route` query argument:
`http://localhost:8888/index.php?rest_route=%2Fwp%2Fv2%2Ftemplates%2Ftt1-blocks%7Cfront-page`

As it turns out, with plain permalinks, the query param is decoded correctly, and the endpoint's regexp matches it as expected.

---

I guess the easy solution is to use a separator character that doesn't get encoded in the URL. 🤔

Reply:
Good catch, maybe we can use `---` separator or something like that? Anyway, I'm finishing my days soon. I can look again tomorrow if no one beats me to it.

Reply:
Yeah, it's not super urgent, and there is an easy workaround to test the PR anyway. 🙂

Reply:
You could also make it `theme/template-name`.

Reply:
@TimothyBJacobs that would be tricky since the theme slug can contain `/` if it's installed in a subdirectory.

Reply:
Well that's a fun feature! It'd be odd, but maybe `//`? Anything else that is URL safe could also appear in the theme name or template path right?

Reply:
I think so yes and I like `//` :)

Reply:
Updated this. I think it works now.